File: selection_sort.js
const fs = require('fs');
const os = require('os');
const process = require('process');

function selectionSort(arr) {
  const n = arr.length;
  for (let i = 0; i < n; i++) {
    let minIdx = i;
    for (let j = i + 1; j < n; j++) {
      if (arr[j] < arr[minIdx]) {
        minIdx = j;
      }
    }
    [arr[i], arr[minIdx]] = [arr[minIdx], arr[i]]; 
  }
}

function main() {
  // Lendo os números do arquivo
  fs.readFile('arq.txt', 'utf8', (err, data) => {
    if (err) {
      console.error('Erro ao abrir o arquivo:', err);
      return;
    }

    const numbers = data.split('\n').map(line => parseInt(line.trim(), 10)).filter(Boolean);

    // Informações do sistema
    console.log("Linguagem: JavaScript (Node.js)");
    console.log("Versão:", process.version);
    console.log("Sistema Operacional:", os.platform());
    console.log("Arquitetura:", os.arch());

    // Monitoramento de desempenho
    const startMemory = process.memoryUsage().rss; // Memória inicial
    const startTime = process.hrtime(); // Tempo inicial

    // Ordenando com Selection Sort
    selectionSort(numbers);

    // Tempo final
    const [seconds, nanoseconds] = process.hrtime(startTime);
    const elapsedTime = (seconds * 1000) + (nanoseconds / 1000000); 

    const endMemory = process.memoryUsage().rss; 

    // Gravando o resultado no arquivo
    fs.writeFile('arq-saida.txt', numbers.join('\n'), 'utf8', (err) => {
      if (err) {
        console.error('Erro ao gravar o arquivo:', err);
        return;
      }

      const memoryUsed = (endMemory - startMemory) / 1024;
      console.log(`Tempo (ms): ${elapsedTime.toFixed(2)}`);
      console.log(`Memória utilizada (KB): ${memoryUsed.toFixed(2)}`);
    });
  });
}

main();

File: Selection_sort.py
import platform
import time
from memory_profiler import memory_usage

def selection_sort(arr):
    n = len(arr)
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

def main():
    # Lendo os números do arquivo
    with open("arq.txt", "r") as file:
        numbers = [int(line.strip()) for line in file]

    # Informações do sistema
    print("Linguagem: Python")
    print("Versão:", platform.python_version())
    print("Sistema Operacional:", platform.system())
    print("Arquitetura:", platform.architecture()[0])

    # Monitoramento de desempenho
    start_time = time.time()
    memory_used = memory_usage((selection_sort, (numbers,)), interval=0.1)[0] * 1024  # Convert to KB

    end_time = time.time()
    elapsed_time = (end_time - start_time) * 1000  # Tempo em milissegundos

    # Gravando o resultado no arquivo
    with open("arq-saida.txt", "w") as file:
        for num in numbers:
            file.write(f"{num}\n")

    print(f"Tempo (ms): {elapsed_time:.2f}")
    print(f"Memória utilizada (KB): {memory_used:.2f}")

if __name__ == "__main__":
    main()
